lib/puppet_x/puppetlabs/powershell_manager.rb

```diff
@@ -0,0 +1,203 @@
+          Process.kill('KILL', @ps_process[:pid])
+        end
+      end
+
+      def template_path
+        File.expand_path('../../templates', __FILE__)
+      end
+
+      def make_ps_code(powershell_code, output_ready_event_name, timeout_ms = 300 * 1000)
+        template_file = File.new(template_path + "/invoke_ps_command.erb").read
+        template = ERB.new(template_file, nil, '-')
+        template.result(binding)
+      end
+
+      private
+
+      def self.is_readable?(stream, timeout = 0.5)
+        read_ready = IO.select([stream], [], [], timeout)
+        read_ready && stream == read_ready[0][0]
+      end
+
+      def self.create_event(name, manual_reset = false, initial_state = false)
+        handle = FFI::Pointer::NULL_HANDLE
+
+        FFI::Pointer.from_string_to_wide_string(name) do |name_ptr|
+          handle = CreateEventW(FFI::Pointer::NULL,
+            manual_reset ? 1 : FFI::WIN32_FALSE,
+            initial_state ? 1 : FFI::WIN32_FALSE,
+            name_ptr)
+
+          if handle == FFI::Pointer::NULL_HANDLE
+            msg = "Failed to create new event #{name}"
+            raise Puppet::Util::Windows::Error.new(msg)
+          end
+        end
+
+        handle
+      end
+
+      WAIT_ABANDONED = 0x00000080
+      WAIT_OBJECT_0 = 0x00000000
+      WAIT_TIMEOUT = 0x00000102
+      WAIT_FAILED = 0xFFFFFFFF
+
+      def self.wait_on(wait_object, timeout_ms = 50)
+        wait_result = Puppet::Util::Windows::Process::WaitForSingleObject(
+          wait_object, timeout_ms)
+
+        case wait_result
+        when WAIT_OBJECT_0
+          Puppet.debug "Wait object signaled"
+        when WAIT_TIMEOUT
+          Puppet.debug "Waited #{timeout_ms} milliseconds..."
+        # only applicable to mutexes - should never happen here
+        when WAIT_ABANDONED
+          msg = 'Catastrophic failure: wait object in inconsistent state'
+          raise Puppet::Util::Windows::Error.new(msg)
+        when WAIT_FAILED
+          msg = 'Catastrophic failure: waiting on object to be signaled'
+          raise Puppet::Util::Windows::Error.new(msg)
+        end
+
+        wait_result
+      end
+
+      def write_stdin(input)
+        @stdin.puts(input)
+      rescue => e
+        msg = "Error writing STDIN / reading STDOUT: #{e}"
+        raise Puppet::Util::Windows::Error.new(msg)
+      end
+
+      def drain_pipe(pipe)
+        output = []
+        while self.class.is_readable?(pipe, 0.1) do
+          l = pipe.gets
+          Puppet.debug "#{Time.now} PIPE> #{l}"
+          output << l
+        end
+        output
+      end
+
+      def read_stdout(output_ready_event, wait_interval_ms = 50)
+        output = []
+        errors = []
+        waited = 0
+
+        # drain the pipe while waiting for the event signal
+        while WAIT_TIMEOUT == self.class.wait_on(output_ready_event, wait_interval_ms)
+          output << drain_pipe(@stdout)
+          errors << drain_pipe(@stderr)
```

Comment:
I'm not sure that this is sufficient to prevent a deadlock - I *think* that these would actually need to alternate and in some way limit the amount of times that the `is_readable?` check is performed.